pkg/maps/vtep_policy/vtep.go

```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package vtep_policy
+
+import (
+	"fmt"
+	"net"
+
+	"log/slog"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/datapath/linux/config/defines"
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/ebpf"
+	"github.com/cilium/cilium/pkg/mac"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/types"
+	"github.com/cilium/hive/cell"
+)
+
+const (
+	MaxEntries = 16384
+	// Name is the canonical name for the VTEP map on the filesystem.
+	VtepPolicyMapName = "cilium_vtep_policy_map"
+)
+
+// Must be in sync with struct vtep_key in <bpf/lib/vtep.h>
+type VtepPolicyKey struct {
+	PrefixLen uint32     `align:"prefixlen"`
+	SourceIP  types.IPv4 `align:"src_ip"`
+	DestCIDR  types.IPv4 `align:"dst_ip"`
+}
+
+func (k VtepPolicyKey) String() string {
+	return fmt.Sprintf("%s.%s/%d", k.SourceIP, k.DestCIDR, k.PrefixLen)
+}
+
+func (k *VtepPolicyKey) New() bpf.MapKey { return &VtepPolicyKey{} }
+
+// NewKey returns an Key based on the provided source IP address and destination CIDR
+func NewKey(srcIP net.IP, dstCIDR *cidr.CIDR) VtepPolicyKey {
+	result := VtepPolicyKey{}
+
+	if ip4 := srcIP.To4(); ip4 != nil {
+		copy(result.SourceIP[:], ip4)
+	}
+
+	if ip4 := dstCIDR.IP.To4(); ip4 != nil {
+		copy(result.DestCIDR[:], ip4)
+	}
+	ones, _ := dstCIDR.IPNet.Mask.Size()
+	result.PrefixLen = 32 + uint32(ones)
+
+	return result
+}
+
+// VtepPolicyVal implements the bpf.MapValue interface. It contains the
+// VTEP endpoint MAC and IP
+type VtepPolicyVal struct {
+	Mac    mac.Uint64MAC `align:"vtep_mac"`
+	VtepIp types.IPv4    `align:"tunnel_endpoint"`
+	_      [4]byte
+}
+
+func (v *VtepPolicyVal) String() string {
+	return fmt.Sprintf("vtepmac=%s tunnelendpoint=%s",
+		v.Mac, v.VtepIp)
+}
+
+func (v *VtepPolicyVal) New() bpf.MapValue { return &VtepPolicyVal{} }
+
+// Map represents an VTEP BPF map.
+type VtepPolicyMap struct {
+	m *bpf.Map
+}
+
+func createPolicyMapFromDaemonConfig(in struct {
+	cell.In
+
+	Lifecycle cell.Lifecycle
+	*option.DaemonConfig
+	MetricsRegistry *metrics.Registry
+}) (out struct {
+	cell.Out
+
+	VtepMap bpf.MapOut[*VtepPolicyMap]
+	defines.NodeOut
+}) {
+	if !in.EnableVTEP {
+		return
+	}
+
+	if in.EnableIPv4 {
+		out.VtepMap = bpf.NewMapOut(newVtepPolicyMap(in.Lifecycle, in.MetricsRegistry, ebpf.PinByName))
+	}
+
+	return
+}
```

Comment:
I think we can simplify, the whole in/out structs are a bit much for the amount of arguments and return values.

```suggestion
func createPolicyMapFromDaemonConfig(
	lifecycle cell.Lifecycle,
	config *option.DaemonConfig,
	metricsRegistry *metrics.Registry,
) bpf.MapOut[*VtepPolicyMap] {
	if !config.EnableVTEP || !config.EnableIPv4 {
		return bpf.NewMapOut[*VtepPolicyMap](nil)
	}

	return bpf.NewMapOut(newVtepPolicyMap(lifecycle, metricsRegistry, ebpf.PinByName))
}
```